flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTimeoutTest.java

```diff
@@ -0,0 +1,200 @@
+    private class TimeoutWriter extends AsyncSinkWriter<String, Long> {
+        private Exception fatalError;
+        private final CountDownLatch completionLatch;
+        private Thread submitThread;
+
+        private boolean shouldFailRequest = false;
+
+        public TimeoutWriter(
+                WriterInitContext writerInitContext,
+                int maxBatchSize,
+                long maximumTimeInBufferMs,
+                long requestTimeout,
+                boolean failOnTimeout) {
+            super(
+                    (ElementConverter<String, Long>) (element, context) -> Long.parseLong(element),
+                    writerInitContext,
+                    AsyncSinkWriterConfiguration.builder()
+                            .setMaxBatchSize(maxBatchSize)
+                            .setMaxBatchSizeInBytes(Long.MAX_VALUE)
+                            .setMaxInFlightRequests(Integer.MAX_VALUE)
+                            .setMaxBufferedRequests(Integer.MAX_VALUE)
+                            .setMaxTimeInBufferMS(maximumTimeInBufferMs)
+                            .setMaxRecordSizeInBytes(Long.MAX_VALUE)
+                            .setRequestTimeoutMS(requestTimeout)
+                            .setFailOnTimeout(failOnTimeout)
+                            .build(),
+                    Collections.emptyList());
+            this.completionLatch = new CountDownLatch(1);
+        }
+
+        @Override
+        protected void submitRequestEntries(
+                List<Long> requestEntries, ResultHandler<Long> resultHandler) {
+            submitThread =
+                    new Thread(
+                            () -> {
+                                while (completionLatch.getCount() > 0) {
+                                    try {
+                                        completionLatch.await();
+                                    } catch (InterruptedException e) {
+                                        fail("Submission thread must not be interrupted.");
+                                    }
+                                }
+                                submitRequestEntriesSync(requestEntries, resultHandler);
+                            });
+
+            submitThread.start();
+        }
+
+        private void submitRequestEntriesSync(
+                List<Long> requestEntries, ResultHandler<Long> resultHandler) {
+            if (fatalError != null) {
+                resultHandler.completeExceptionally(fatalError);
+            } else if (shouldFailRequest) {
+                shouldFailRequest = false;
+                resultHandler.retryForEntries(requestEntries);
+            } else {
+                destination.addAll(requestEntries);
+                resultHandler.complete();
+            }
+        }
+
+        @Override
+        protected long getSizeInBytes(Long requestEntry) {
+            return 8;
+        }
+
+        public void setFatalError(Exception fatalError) {
```

Comment:
Should be ensured thread-safe modification

Reply:
Only modified by test thread